code/components/extra-natives-five/src/WaterExtraNatives.cpp

```diff
@@ -0,0 +1,148 @@
+{
+	static void* (*fn)(__m128*, bool, uint64_t) = (decltype(fn))g_testForWaterVisualsAddr;
+	return fn(entityCoords, a2, a3);
+}
+
+void* _TestForWaterPhysics(void* a1, __m128* entityCoords, uint64_t a3) //vanilla version
+{
+	static void* (*fn)(void*, __m128*, uint64_t) = (decltype(fn))g_testForWaterPhysicsAddr;
+	return fn(a1, entityCoords, a3);
+}
+bool ShouldVolumeBeDry(__m128* entityCoords)
+{
+	bool result = false;
+	for (const sDepthBound& area : g_dryAreas)
+	{
+		result += IsInBounds(entityCoords, area);
+	};
+	return result;
+}
+
+void* CustomTestForWaterPhysics(void* a1, __m128* entityCoords, uint64_t a3)
+{
+	if (ShouldVolumeBeDry(entityCoords))
+	{
+		return 0;
+	}
+	return _TestForWaterPhysics(a1, entityCoords, a3);
+}
+
+void* CustomTestForUnderwaterVisuals(__m128* entityCoords, bool a2, uint64_t a3)
+{
+	if (ShouldVolumeBeDry(entityCoords))
+	{
+		return 0;
+	}
+	return _TestForUnderWaterVisuals(entityCoords, a2, a3);
+}
+
+static HookFunction initFunction([]()
+{
+	g_testForWaterPhysicsAddr = hook::get_pattern("48 8B C4 55 53 56 57 41 54 41 56 41 57 48 8D A8 ? ? ? ? 48 81 EC ? ? ? ? F3 0F 10 42 ? F3 0F 10 0A F3 0F 10 52 ? 0F 29 70 B8 0F 29 78 A8 44 0F 29 40 ? 45 8A F0");
+	g_testForWaterVisualsAddr = hook::get_pattern("48 8B C4 48 89 58 08 48 89 68 18 48 89 70 20 57 48 81 EC ? ? ? ? 83 48 10 FF");
+
+	//Physics calls
+	hook::call(hook::get_pattern("E8 ? ? ? ? 48 8B CB E8 ? ? ? ? 44 38 3D ? ? ? ? 74 10 8A 83"), CustomTestForWaterPhysics);
+	hook::call(hook::get_pattern("E8 ? ? ? ? EB 07 83 A1 ? ? ? ? ? 48 83 C4 38 C3"), CustomTestForWaterPhysics);
+	hook::call(hook::get_pattern("E8 ? ? ? ? 0F 28 CE 48 8B CF E8 ? ? ? ? 66 89 B7 ? ? ? ? 89 B7 ? ? ? ? 8A 8F ? ? ? ? 3A 8F"), CustomTestForWaterPhysics);
+
+	//Visual calls
+	hook::call(hook::get_pattern("E8 ? ? ? ? 33 F6 84 C0 0F 84 ? ? ? ? F3 0F 10 4F ? F3 0F 10 07 E8"), CustomTestForUnderwaterVisuals);
+
+	fx::ScriptEngine::RegisterNativeHandler("CREATE_DRY_VOLUME", [=](fx::ScriptContext& context)
+	{
+		auto newVolume = new sDepthBound(
+		new float[]{
+		context.GetArgument<float>(0),
+		context.GetArgument<float>(1),
+		context.GetArgument<float>(2),
+		},
+		new float[]{
+		context.GetArgument<float>(3),
+		context.GetArgument<float>(4),
+		context.GetArgument<float>(5),
+		});
+
+		std::vector<sDepthBound>::iterator v = std::find(g_dryAreas.begin(), g_dryAreas.end(), newVolume);
+		if (v != g_dryAreas.end())
+		{
+			int index = v - g_dryAreas.begin();
+			context.SetResult<int>(index);
+			trace("A duplicate dry volume was attempted to be created, the existing handle has been returned instead.\n");
+		}
+		else
+		{
+			context.SetResult<int>(g_dryAreas.size());
+			g_dryAreas.push_back(*newVolume);		
+		}
+	});
+
+	fx::ScriptEngine::RegisterNativeHandler("REMOVE_DRY_VOLUME", [=](fx::ScriptContext& context)
+	{
+		int index = context.GetArgument<int>(0);
+		if (index > g_dryAreas.size() || index < 0)
```

Comment:
I believe there should be `index >= g_dryAreas.size()`.
You can remove dry volume indexed as `0` when you actually have no dry volumes created, this will cause client crash.
When you have at least 1 dry volume, if you call `REMOVE_DRY_VOLUME` for next index (i.e. for next volume, when it wasn't created yet) and then call `CREATE_DRY_VOLUME`, you will get crash.